README.md
# Currency Converter 

A currency converter GUI app developed with python and tkinter.

This application is created by me to convert currencies by inputting two 
currency codes and clicking the "Convert now" button. 

This application used the Exchange Rates API to receive the converted result and 
users are able to manipulate the output result by inputting a certain amount to be converted,
and set a historical date. 

The user can click on the history button to check their past pairs of currency code converted that are in a 
form of a button in another GUI and they are able to click on one of the buttons to be input the pairs of codes 
into the main GUI again to be used as conversion.

The user can also click on the date checkmark to have a historical rate when converting currencies. The user can also 
click the "Set date" button to set a past date for a historical rate they want in another GUI.  

This app contains a history.txt file to save and track the user past history of any pair of currency they have converted with. 
The pair of currencies code are written as ("from-currency-code"-"to_currency_code") in the txt file. User should not put any other format 
in this txt file or it will break the application. When the user close the main GUI, the program will then write their new history of converted pairs of currency code 
into the txt file. 

This projects imports the tkcalendar and requests modules and you will need to install both of 
these modules to use this GUI application. 

history.txt
USD-MYR
MYR-USD
AUD-HKD
HKD-CAD
CAD-HKD


main.py
# Importing modules
from tkinter import *
import converter_GUI
from converter_GUI import GUI
from queue_ds import history_queue

# main function to execute all of the code
def main():
    # queue data structure
    history_list = history_queue()

    #main GUI
    root = Tk()
    root.title("Currency Converter")
    root.resizable(0, 0)
    root.geometry(f"{converter_GUI.width}x{converter_GUI.height}")
    root.iconbitmap("currency.ico")
    root.protocol("WM_DELETE_WINDOW", lambda: converter_GUI.main_closing(root, history_list))

    # context manager to read the history.txt file and put them into a queue
    with open("history.txt", "r") as file:
        for line in file.readlines():
            if "\n" in line:
                history_list.queue.append(line[:-1])
            else:
                history_list.queue.append(line)
        if history_list.queue[-1] == "":
            history_list.queue.pop()

    # creating 3 frames one for date and history, one for convert button and another for setting input
    # top left frame for setting input
    top_frame = Frame(root,
                      bg="#add8e6",
                      width=converter_GUI.prct(72, converter_GUI.width),
                      height=converter_GUI.prct(70, converter_GUI.height))
    top_frame.place(x = 0, y = 0)

    # right frame for date and history
    right_frame = Frame(root,
                        bg="#90ee90",
                        width=converter_GUI.prct(28, converter_GUI.width),
                        height=converter_GUI.height)
    right_frame.place(x = converter_GUI.prct(72, converter_GUI.width) , y = 0)

    # bottom frame for convert button
    bottom_frame = Frame(root,
                         bg="#808080",
                         width=converter_GUI.prct(72, converter_GUI.width),
                         height=converter_GUI.prct(30, converter_GUI.height))
    bottom_frame.place(x=0, y=converter_GUI.prct(70, converter_GUI.height))

    # label text
    insert_code_text = Label(top_frame, text="Insert currency code", font=("Verdana bold", 20), bg="#add8e6").place(
        x=converter_GUI.prct(18, converter_GUI.width), y=0)
    amount_text = Label(top_frame, text="Amount", font=("Verdana bold", 18), bg="#add8e6").place(
        x=converter_GUI.prct(28, converter_GUI.width), y=converter_GUI.prct(42, converter_GUI.height))

    # Entry
    GUI.create_entry(top_frame)
    GUI.create_convert_button(bottom_frame, history_list)

    # history stuff
    GUI.create_history_btn(right_frame, history_list)
    GUI.create_date_settings(right_frame)

    mainloop()


if __name__ == '__main__':
    main()


btn_cmd.py
"""This module contains staticmethod from the btn_commands class that are used as commands for the buttons in the GUI"""
# import modules
from Converter_api import api_convert
from tkinter import *
from tkinter import messagebox
from tkcalendar import *
from datetime import date
from history_buttons import array_btns, on_closing

# class method that are used as commands by the buttons in the GUI
class btn_commands:
    today_date = date.today()

    # static method for the swap button command
    @staticmethod
    def swap(cls):
        current_from_code, current_to_code = cls.from_code.get(), cls.to_code.get()
        if cls.from_amount.get() != "" and cls.to_amount.get() != "":
            current_from_amount, current_to_amount = cls.from_amount.get(), cls.to_amount.get()
            cls.from_amount.set(current_to_amount)
            cls.to_amount.set(current_from_amount)
        cls.from_code.set(current_to_code)
        cls.to_code.set(current_from_code)

    # static method for the convert button command
    @staticmethod
    def start_convert(lft_code, rgt_code, lft_amt, rgt_amt, date_check, date, queue):
        if lft_code == "" or rgt_code == "":
            messagebox.showerror(title="Error! Boxes are not fill",
                                 message="Please fill in the two top boxes with currency codes")
        elif float(lft_amt.get()) == 0:
            messagebox.showerror(title="Error! Amount input is zero",
                                 message="Please fill in the left bottom box with a value that is not 0")
        else:
            # setting date to false if date check value is 0 else it is put as its default string
            date = date if date_check else False
            converted_result = api_convert(lft_code, rgt_code, lft_amt.get(), date)
            if converted_result is None:
                messagebox.showerror(title="Error! User inputs are not currency code",
                                     message="Please fill in the two top boxes with currency codes")
            else:
                rgt_amt.set(converted_result)
            queue.modify(f"{lft_code}-{rgt_code}")

    # static method for the history button command
    @staticmethod
    def history(btn, lft_var, rgt_var, queue):
        btn["state"] = "disabled" # disabling the button so that it doesn't make the user open more than one
        history_GUI = Tk()
        history_GUI.resizable(0, 0)
        history_GUI.geometry("160x280")
        history_GUI.protocol("WM_DELETE_WINDOW", lambda: on_closing(btn, history_GUI))
        buttons_list = array_btns(history_GUI, queue.queue, lft_var, rgt_var, btn)

    # static method for the set button command for date settings
    @staticmethod
    def set(btn, set_date, label):
        btn["state"] = "disabled" # disabling the button so that it doesn't make the user open more than one
        global date_GUI
        date_GUI = Tk()
        date_GUI.title("Date settings")
        date_GUI.resizable(0, 0)
        date_GUI.geometry("250x250")
        date_GUI.protocol("WM_DELETE_WINDOW", lambda: on_closing(btn, date_GUI))

        # stringvar for calendar
        date_var = StringVar(date_GUI, value=f"{29}-{7}-{2022}")
        # calendar class from tkcalendar
        cal = Calendar(date_GUI, selectmode="day", textvariable=date_var, date_pattern="y-mm-dd")
        date_var.set(set_date.get()) # change the variable to the default isoformat that is needed for the api
        cal.place(x=0, y=0)
        date_label = Label(date_GUI, textvariable=date_var, font="Times 15").place(x=80, y=190)
        set_btn = Button(date_GUI, text="Set",
                         command=lambda: btn_commands.confirm_date(date_var.get(),set_date, on_closing(btn, date_GUI)))
        set_btn.place(x=125, y=225)


    # static method for setting date button in date GUI
    @staticmethod
    def confirm_date(new_date, current_date, function):
        if date.fromisoformat(new_date) > btn_commands.today_date:
            messagebox.showerror(title="Error! Date is a future date!",
                                 message="Please select a date that is not set in the future of the today's date")
        else:
            current_date.set(new_date)
            return function
            # close gui after setting by returning the on closing function



history_buttons.py
"""This module is used for the history GUI to create the buttons"""
# import modules
from tkinter import Button

# array class to hold all of the buttons in the history GUI
class array_btns:
    def __init__(self, master, arr, left_var, right_var, btn):
        self.btns_list = list()
        for element in arr:
            button = his_btn(master, element, left_var, right_var, btn)
            self.btns_list.append(button)

# button class to hold a pair of code
class his_btn():
    def __init__(self, master, element, left_var, right_var, btn):
        self.button = Button(master, text=f"{element}", width=20,
                             command=lambda:his_btn.select(element, master, left_var, right_var, btn))
        self.button.pack()

    # static method for the button select command
    @staticmethod
    def select(element, master, left_var, right_var, btn):
        left_code, right_code = element.split("-")
        left_var.set(left_code)
        right_var.set(right_code)
        return on_closing(btn, master)

# on closing function for date set and history button protocol
def on_closing(btn, GUI):
    try:
        btn["state"] = "normal"
        GUI.destroy()
    except Exception as e: # if the main GUI is closed, this is used to allow the date gui or history gui to close too
        GUI.destroy()




converter_GUI.py
# Importing modules
from tkinter import *
from datetime import date
from btn_cmd import btn_commands
from threading import Thread

# variables
width = 900
height = 500

# function to measure the width or height of the gui to be used for size or placement of frames or buttons
def prct(prct, length):
    return (prct * length) // 100

class GUI():
    # Class variable
    date = None
    date_check = None
    from_code = None
    to_code = None
    from_amount = None
    to_amount = None

    # static method to create entries
    @staticmethod
    def create_entry(location):
        # stringvar for code Entry
        GUI.from_code = StringVar(location)
        GUI.to_code  = StringVar(location)
        GUI.from_code.trace("w", lambda name, index, mode: callback(GUI.from_code))
        GUI.to_code.trace("w", lambda name, index, mode: callback(GUI.to_code))

        # stringvar for amount Entry, the value will then convert into a float data type to be used in the api
        GUI.from_amount = StringVar(location, value="1.0")
        GUI.from_amount.trace("w", lambda name, index, mode: callback(GUI.from_amount, GUI.to_amount, 7, left_amount))
        GUI.to_amount = StringVar(location)

        # currency code Input
        left_code = Entry(location,width=8, font=("Times 30"), bd=4, textvariable=GUI.from_code)
        left_code.place(x=prct(7.2, width), y=prct(20, height))
        right_code = Entry(location,width=8, font=("Times 30"), bd=4, textvariable=GUI.to_code)
        right_code.place(x=prct(43.2, width), y=prct(20, height))

        # amount Input
        left_amount = Entry(location, width=8, font=("Times 30"), textvariable=GUI.from_amount)
        left_amount.bind("." , lambda e: "break")
        left_amount.place(x=prct(7.2, width), y=prct(40, height))
        right_amount = Entry(location, width=8, font=("Times 30"), textvariable=GUI.to_amount)
        # make the right amount read only
        right_amount.configure(state='disabled')
        right_amount.place(x=prct(43.2, width), y=prct(40, height))

        # buttons
        swap_btn = Button(location, width=5, text="<-\n->", command=lambda:btn_commands.swap(GUI))
        swap_btn.place(x=prct(32, width), y=prct(22, height))

    # static method to create conver button
    @staticmethod
    def create_convert_button(location, queue):
        convert_btn = Button(location, text="Convert", width=8, font=("Verdana bold", 16),
                             command=lambda: Thread(target=btn_commands.start_convert , daemon=True,
                                                    args=(GUI.from_code.get(), GUI.to_code.get(), GUI.from_amount,
                                                          GUI.to_amount, GUI.date_check.get() ,GUI.date.get(), queue)).start())
        convert_btn.place(x=prct(28, width), y=50)

    # static method to create date settings
    @staticmethod
    def create_date_settings(location):
        # date check intvar
        GUI.date_check = IntVar()

        # Date checkmark
        date_checkmark = Checkbutton(location, text="Date check", variable=GUI.date_check,
                                     bg="#90ee90", font=("Times 15"))
        date_checkmark.deselect()  # to turn off the button first when it first runs
        date_checkmark.place(x=prct(6, width),
                             y=prct(55, height))

        # settings date to the current date when the GUI start
        GUI.date = StringVar(location, value=date.today())
        # date text
        date_txt = Label(location, textvariable=GUI.date, font=("Verdana bold", 18))
        date_txt.place(x=prct(4, width), y=prct(68, height))

        setdate_btn = Button(location, text="Set Date", font=("Times 12"),
                             command=lambda: btn_commands.set(setdate_btn,GUI.date, date_txt))
        setdate_btn.place(x=prct(10, width), y=prct(78, height))

    # method to create history button and settings
    @staticmethod
    def create_history_btn(location, queue):
        history_btn = Button(location, text="History", width=8, font=("Times 14"),
                             command=lambda: btn_commands.history(history_btn, GUI.from_code,
                                                                  GUI.to_code, queue))
        history_btn.place(x=prct(8, width), y=prct(20, height))

# function for stringvar trace method
def callback(var, rightvar=None, length=None, entry=None ):
    # this is used for code variables
    if length is None:
        limit = var.get()[0:3].upper()
        var.set(limit)
    else: # this is used for amount variable
        rightvar.set("")
        if len(var.get()) > length:
            limit = var.get()[0:length]
            var.set(limit)
        elif var.get() == "" or var.get() == ".":
            var.set(1.0)
        else:
            limit = " ".join(var.get()).split(" ")
            new_limit = list(filter(lambda x: x.isdigit() or x == ".", limit ))
            if new_limit.count(".") != 1:
                entry.unbind(".")
            else:
                entry.bind("." , lambda e: "break")
            var.set("".join(new_limit))

# function for main GUI protocol
def main_closing(main_GUI, his_list):
    with open("history.txt", "w") as file:
        for code in his_list.queue:
            file.write(code)
            file.write("\n")
    main_GUI.destroy()

